Add Starlink Integration

boswelja opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
Add support for loading data from a Starlink terminal. I have tested the integration with my square Dishy & new terminal, and the underlying python library `starlink-grpc-core` has been used for both (new) square Dishy and (old) round Dishy.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/23993

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

homeassistant commented:
Hi @boswelja, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#77091).

Once you do that we will be able to review and accept this pull request.

Thanks!

boswelja commented:
I am working on tests now, I'd appreciate review on the code that's there in case there's something significant I've missed :)

frenck commented on homeassistant/components/starlink/__init__.py:
```diff
@@ -0,0 +1,36 @@
+from .const import CONFIG_POLL_RATE, DOMAIN
+from .coordinator import StarlinkUpdateCoordinator
+
+PLATFORMS: list[Platform] = [Platform.SENSOR, Platform.BINARY_SENSOR, Platform.BUTTON]
```
Please limit the initial PR to a single platform. See also:

<https://developers.home-assistant.io/docs/creating_component_code_review/#5-make-your-pull-request-as-small-as-possible>

frenck commented on homeassistant/components/starlink/__init__.py:
```diff
@@ -0,0 +1,36 @@
+    """Set up Starlink from a config entry."""
+    coordinator = StarlinkUpdateCoordinator(
+        hass=hass,
+        url=entry.data[CONF_URL],
+        poll_rate=entry.data[CONFIG_POLL_RATE],
+        name=entry.title,
```
Alternatively, pass the full ConfigEntry to the coordinator.

boswelja replied:
Is that the preferred approach? I'm relatively new to Python but I figure having a class expect things like URL instead of taking the entire config would be preferred

boswelja replied:
I think this is less relevant now, since the coordinator takes less data. Please re-open if you disagree 🙂

frenck commented on homeassistant/components/starlink/__init__.py:
```diff
@@ -0,0 +1,36 @@
+
+    hass.data[DOMAIN] = {entry.entry_id: coordinator}
+
+    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
```
```suggestion
    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
```

frenck commented on homeassistant/components/starlink/binary_sensor.py:
```diff
@@ -0,0 +1,50 @@
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from ...helpers.entity import DeviceInfo
```
Don't use relative paths outside your own integration package.

boswelja replied:
Good catch! I've noticed sometimes VS Code will use relative imports for homeassistant modules, I wonder why it chose to do it here

frenck commented on homeassistant/components/starlink/config_flow.py:
```diff
@@ -0,0 +1,77 @@
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_NAME, default="Starlink"): str,
```
Either extract the name from the device, or give it a default name. We like to keep the number of options/questions in a configuration setup flow as low as possible.

The suggestion is to remove the name field

frenck commented on homeassistant/components/starlink/config_flow.py:
```diff
@@ -0,0 +1,77 @@
+    {
+        vol.Required(CONF_NAME, default="Starlink"): str,
+        vol.Required(CONF_URL, default="192.168.100.1:9200"): str,
+        vol.Required(CONFIG_POLL_RATE, default=1): vol.All(
+            vol.Coerce(int), vol.Range(min=1)
+        ),
```
Integrations should have a default/sane polling interval. If a user wants a customized polling interval, the `homeassistant.update_entity` service can be used in combination with an automation to create any update schedule they can imaging.

please remove this option.

frenck commented on homeassistant/components/starlink/config_flow.py:
```diff
@@ -0,0 +1,77 @@
+        vol.Required(CONFIG_POLL_RATE, default=1): vol.All(
+            vol.Coerce(int), vol.Range(min=1)
+        ),
+        vol.Required(
+            CONFIG_LINK_SPEED_FORMAT, default=DATA_RATE_MEGABITS_PER_SECOND
+        ): vol.In(
+            [
+                DATA_RATE_BITS_PER_SECOND,
+                DATA_RATE_MEGABITS_PER_SECOND,
+                DATA_RATE_GIGABITS_PER_SECOND,
+            ]
+        ),
```
Can we detect this? Is this needed?

boswelja replied:
The default link speed format is bit/s, which I don't find particularly useful. That said, I also didn't want to lock users out of betting the raw data, so I figured a config option would be best.
If you would still like it removed, I can just lock it to Mbit/s

frenck commented on homeassistant/components/starlink/config_flow.py:
```diff
@@ -0,0 +1,77 @@
+        errors = {}
+        if user_input is not None:
+            # Input validation
+            url = user_input[CONF_URL]
+            name = user_input[CONF_NAME]
+            poll_rate = user_input[CONFIG_POLL_RATE]
```
```suggestion
```

frenck commented on homeassistant/components/starlink/config_flow.py:
```diff
@@ -0,0 +1,77 @@
+            url = user_input[CONF_URL]
+            name = user_input[CONF_NAME]
+            poll_rate = user_input[CONFIG_POLL_RATE]
+            uid = await self.get_device_id(url=url)
```
Can the `uid` be used as unique identifier for the configuration flow?

frenck replied:
Won't this throw/raise errors when the connection fails?

boswelja replied:
The error is caught by `get_device_id` and `None` is returned, indicating the device doesn't exist. I have tested this manually by inputting a non-Starlink terminal IP address :)

frenck commented on homeassistant/components/starlink/config_flow.py:
```diff
@@ -0,0 +1,77 @@
+            url = user_input[CONF_URL]
+            name = user_input[CONF_NAME]
+            poll_rate = user_input[CONFIG_POLL_RATE]
+            uid = await self.get_device_id(url=url)
+            # If everything looks good, create the entry
+            if uid is not None and poll_rate > 0:
```
```suggestion
            if await self.get_device_id(url=user_input[CONF_URL]):
```

boswelja replied:
I have kept the separate variable to pass to `async_set_unique_id`

frenck commented on homeassistant/components/starlink/config_flow.py:
```diff
@@ -0,0 +1,77 @@
+            if uid is not None and poll_rate > 0:
+                return self.async_create_entry(
+                    title=name,
+                    data={
+                        CONF_URL: url,
+                        CONFIG_POLL_RATE: poll_rate,
+                        CONFIG_LINK_SPEED_FORMAT: user_input[CONFIG_LINK_SPEED_FORMAT],
+                    },
```
```suggestion
                    data=user_input,
```

frenck commented on homeassistant/components/starlink/config_flow.py:
```diff
@@ -0,0 +1,77 @@
+            step_id="user", data_schema=CONFIG_SCHEMA, errors=errors
+        )
+
+    async def get_device_id(self, url: str) -> str | None:
+        """Get the device UID."""
+        try:
+            context = ChannelContext(target=url)
+            response = await self.hass.async_add_executor_job(lambda: get_id(context))
+            return response
+        except GrpcError:
+            return None
```
Maybe just inline this in the config flow above?

boswelja replied:
Ideally this will be used for auto-discovery (since the typical Starlink setup has Dishy on a fixed IP address), I just haven't figured out how to do that yet

frenck commented on homeassistant/components/starlink/coordinator.py:
```diff
@@ -0,0 +1,89 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+DISHY_ID = "dishy_id"
+DISHY_HARDWARE = "dishy_hardware"
```
Maybe move these to the constants?

frenck commented on homeassistant/components/starlink/coordinator.py:
```diff
@@ -0,0 +1,89 @@
+            ),
+        )
+
+    def get_device(self) -> DeviceInfo:
+        """Get a DeviceInfo for Dishy."""
+        config_url = "http://" + self.channel_context.target.split(":")[0]
+        return DeviceInfo(
+            identifiers={
+                (DISHY_ID, self.data.id),
+                (DISHY_HARDWARE, self.data.hardware_version),
+            },
+            sw_version=self.data.software_version,
+            hw_version=self.data.hardware_version,
+            name=self.name,
+            configuration_url=config_url,
+            manufacturer="SpaceX",
+            model="Starlink",
+        )
```
This could be part of a base entity instead (create an `entity.py` file and set the device info in that base entity. Extend all other platform entities from this case entity).

boswelja replied:
I like that idea!

frenck commented on homeassistant/components/starlink/coordinator.py:
```diff
@@ -0,0 +1,89 @@
+                return dishy_status
+            raise UpdateFailed
+
+    async def _get_status_data(self) -> DishyStatus | None:
```
This can be inlined into the update data function above, it is what it does.

frenck commented on homeassistant/components/starlink/dish_status.py:
```diff
@@ -0,0 +1,21 @@
+
+
+@dataclass
+class DishyStatus:
```
I'm not entirely sure why this dataclass is needed? `get_status` already seems to return an object that can be used/passed instead?

boswelja replied:
It does return an object, but there's no type hints (yet). dataclass is there just so I get nice autocomplete outside coordinator

frenck commented on homeassistant/components/starlink/strings.json:
```diff
@@ -0,0 +1,13 @@
+      }
+    },
+    "abort": {
+      "single_instance_allowed": "[%key:common::config_flow::abort::single_instance_allowed%]",
```
Doesn't seem to be used?

frenck commented on homeassistant/components/starlink/strings.json:
```diff
@@ -0,0 +1,13 @@
+{
+  "config": {
+    "step": {
+      "confirm": {
```
I don't see a confirm step?

translations seem to be missing

frenck commented on homeassistant/components/starlink/translations/en.json:
```diff
@@ -0,0 +1,28 @@
+{
```
`string.json` is a source file for specific translations like this `en.json` file, yet, they seem to be out of sync.

boswelja replied:
I see, have you got a link to any relevant documentation? I wasn't able to find much in the integration guide re. translations

boswelja replied:
I have made an attempt to update strings, please let me know if I got anything wrong

frenck reviewed (dismissed):
Hi there @boswelja 👋 

Some initial comments to help on getting this PR moving 👍 

../Frenck

avee87 commented on homeassistant/components/starlink/sensor.py:
```diff
@@ -0,0 +1,205 @@
+    )
+
+
+class PingSensor(StarlinkSensorEntity):
```
Did you consider using `SensorEntityDescription` pattern for these sensors? Would probably avoid creating a class for each sensor.

boswelja replied:
I didn't even know it existed 😆
I'll look into it

boswelja replied:
I had a quick look at SensorEntityDescription, and I can't see a method of telling the entity how to load data, so I'm not sure I can get rid of all the classes. Am I missing something?

avee87 replied:
You can extend it with custom fields - for example, add a getter for state. See example here: https://github.com/home-assistant/core/blob/dev/homeassistant/components/wled/sensor.py#L34-L47

boswelja replied:
I have made changes, please let me know what you think :)

avee87 commented on homeassistant/components/starlink/sensor.py:
```diff
@@ -0,0 +1,205 @@
+        super().__init__(coordinator)
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
```
Since you use `CoordinatorEntity` you don't need to override `_handle_coordinator_update()`. You can override `native_value()` - it will be called automatically when coordinator updates.

avee87 commented on homeassistant/components/starlink/sensor.py:
```diff
@@ -0,0 +1,205 @@
+        self, coordinator: StarlinkUpdateCoordinator, config_name: str
+    ) -> None:
+        """Initialize the sensor and set the update coordinator."""
+        self._attr_name = config_name + " Ping"
```
New entity naming is required for all new integrations: https://developers.home-assistant.io/docs/core/entity#has_entity_name-true-mandatory-for-new-integrations
Essentially, you want this to be just `Ping` and use `config_name` for device name. Also, set `_attr_has_entity_name = True`

boswelja replied:
Ah I was wondering if there was something like this, thanks for pointing it out!

avee87 commented on homeassistant/components/starlink/sensor.py:
```diff
@@ -0,0 +1,125 @@
+    _attr_state_class = SensorStateClass.MEASUREMENT
+    _attr_native_unit_of_measurement = TIME_MILLISECONDS
+    _attr_icon = "mdi:speedometer"
+    _attr_unique_id = "ping"
```
I would suggest not using attributes of a base class for non-final values. Basically, when you see `_attr_unique_id` you expect it to be THE unique ID of an entity, not something that is used to construct it somewhere else.
Again, if you use EntityDescription this will become `entity_description.key`

avee87 commented on homeassistant/components/starlink/sensor.py:
```diff
@@ -0,0 +1,125 @@
+
+    _attr_device_class = SensorDeviceClass.TIMESTAMP
+    _attr_unique_id = "last_boot_time"
+    _attr_name = "Last Boot Time"
```
Since you use new naming scheme, you should use sentence case:

> Entity names should start with a capital letter, the rest of the words are lower case (unless it's a proper noun or a capitalized abbreviation of course)

boswelja commented:
If we're happy with the state of the config flow and sensors, I will begin writing some tests

avee87 commented on homeassistant/components/starlink/entity.py:
```diff
@@ -0,0 +1,69 @@
+    """Describes a Starlink sensor entity."""
+
+
+class StarlinkSensorEntity(CoordinatorEntity[StarlinkUpdateCoordinator], SensorEntity):
```
I think most integrations would have all contents of this file in `sensor.py` although I don't think it's a hard requirement.

boswelja replied:
I plan on having more platforms after this PR, I figure it makes sense to store base entity classes all in one place 😄

avee87 commented on homeassistant/components/starlink/entity.py:
```diff
@@ -0,0 +1,69 @@
+    @property
+    def device_info(self) -> DeviceInfo | None:
+        """Get a DeviceInfo for Dishy."""
+        config_url = "http://" + self.coordinator.channel_context.target.split(":")[0]
```
Let's use f-string here

avee87 commented on homeassistant/components/starlink/coordinator.py:
```diff
@@ -0,0 +1,61 @@
+            hass,
+            _LOGGER,
+            name=name,
+            update_interval=timedelta(seconds=1),
```
Do you actually need to poll every second? Looking at exposed sensors I don't think it makes sense to update any of them more than once per minute. Maybe even less often

boswelja replied:
1 second matches the Starlink app, any longer and we risk missing events for more critical sensors such as obstruction status (which aren't in this PR because that's a different platform). If you have a better solution, I'm all ears 🙂

boswelja replied:
Closing this as no better solutions have been provided

lyricnz replied:
Regardless, it should be user configurable, right?

avee87 commented on homeassistant/components/starlink/coordinator.py:
```diff
@@ -0,0 +1,61 @@
+            _LOGGER,
+            name=name,
+            update_interval=timedelta(seconds=1),
+            request_refresh_debouncer=Debouncer(
```
This would probably not be needed with longer update intervals

boswelja replied:
The debouncer probably isn't enev needed now, it was implemented to work around a flaky API Buti since changed how the integration fetches data. I'll remove it

avee87 reviewed (commented):
Sensors look good to me - see some comments though.
Not sure about config flow, I don't know much about it.

avee87 commented on homeassistant/components/starlink/sensor.py:
```diff
@@ -0,0 +1,77 @@
+        key="last_boot_time",
+        name="Last boot time",
+        icon="mdi:download",
+        state_class=SensorStateClass.MEASUREMENT,
```
I believe, this sensor does not fit the [definition of measurement](https://developers.home-assistant.io/docs/core/entity/sensor/#available-state-classes) and long term stats don't make sense for it. Let's remove this line.

boswelja replied:
Good catch! Looks like I missed some stuff while copy-pasting

boswelja commented on tests/components/starlink/test_config_flow.py:
```diff
@@ -0,0 +1,52 @@
+    assert result["errors"]
+
+
+async def test_flow_user_success(hass: HomeAssistant) -> None:
```
I'm not sure what's wrong here, but the second test always fails.
For example, if I run the test as-is now, `test_flow_user_success` will fail. If I swap the order of the tests, then it will be `test_flow_user_fails_no_dishy` will be the one failing.
The failure is from `assert result["type"]`, where the type of the second test matches the first test (thus the assertion fails)
If anyone has any ideas, please let me know

boswelja replied:
Fixed, I ended up patching the `get_device_id` in the config flow class

boswelja committed:
Create Starlink integration

boswelja committed:
Create Starlink integration

boswelja commented:
@frenck Can you please remove the `docs-missing` tag? I have since created and linked a documentation PR

MrConorAE commented:
Hi!
I have a original round dishy and am keen to test out this integration. Is there any way to download and install it in it's current state?

Thanks,
MrConorAE

boswelja commented:
> Hi!
> I have a original round dishy and am keen to test out this integration. Is there any way to download and install it in it's current state?
> 
> Thanks,
> MrConorAE

You could convert it to a custom component, I think you only need to add a version to the manifest but I could be wrong. 

Alternatively you can check out the underlying library, which provides scripts that pull data from Dishy and output to console/database/file

boswelja committed:
Update starlink-grpc-core

boswelja committed:
Update starlink-grpc-core

boswelja commented:
re. integration quality scale, does this qualify for silver? The only thing I'm not 100% sure of is handling device unavailable, but I think coordinator does that for us

boswelja committed:
Run hassfest

archerne commented:
> Hi! I have a original round dishy and am keen to test out this integration. Is there any way to download and install it in it's current state?
> 
> Thanks, MrConorAE

MrConorAE, I have a custom component that I have been working on for the last couple months you can try. https://community.home-assistant.io/t/starlink-dish-and-network-stats/297155/15

MrConorAE commented:
I've been using it already - it's great! Thanks :)

lyricnz commented:
Keen to see this

lyricnz commented:
Testing locally here, seems to be working OK (happy to help, in roughly your timezone, I'm a python dev, just noob with HA).

Are you planning on adding more sensor/buttonss like the integration from @archerne ?

boswelja commented:
Yes, more platforms are planned. As per the PR requirements, we can only add one platform for the initial PR so this needs to be merged first 😄

bdraco commented:
> * Link to documentation pull request:

Be sure to fill this out above

boswelja committed:
Bump GRPCIO to support new integration

boswelja committed:
Bump GRPCIO to support new integration

boswelja commented:
> > * Link to documentation pull request:
> 
> Be sure to fill this out above

Thanks for pointing that out! Can't believe I missed it 😅

boswelja commented:
@bdraco I'd appreciate some direction on these GRPCIO conflicts, they seem to have turned into protobuf conflicts but I cannot find a version to update

bdraco commented:
`protobuf` is hard pinned to a 3.x version for `esphome` and `cast`

There currently isn't a path forward to upgrade to the 4.x series yet

lyricnz commented:
FWIW I think we need a way to ease off the 1-second polling of starlink.  My rpi is wifi and gets not-infrequent:
```
2022-10-15 10:34:50.000 ERROR (MainThread) [homeassistant.components.starlink.coordinator] Timeout fetching Starlink data
2022-10-15 10:34:53.089 INFO (MainThread) [homeassistant.components.starlink.coordinator] Fetching Starlink data recovered
```

boswelja commented:
> FWIW I think we need a way to ease off the 1-second polling of starlink. My rpi is wifi and gets not-infrequent:
> 
> ```
> 2022-10-15 10:34:50.000 ERROR (MainThread) [homeassistant.components.starlink.coordinator] Timeout fetching Starlink data
> 2022-10-15 10:34:53.089 INFO (MainThread) [homeassistant.components.starlink.coordinator] Fetching Starlink data recovered
> ```
> 
> Almost none of the metrics need that level of frequency (only, maybe, "obstructed").

Thanks for the feedback! I originally had the polling rate configurable, but was advised it should not be. I'm open to suggestions :)

boswelja commented:
> 



> `protobuf` is hard pinned to a 3.x version for `esphome` and `cast`
> 
> There currently isn't a path forward to upgrade to the 4.x series yet

That does not sound good. I'm not familiar with Python, but is there a way we can mix versions or suppress the warning? This appears to run locally without fiddling with GRPCIO versions

boswelja committed:
Revert "Bump GRPCIO to support new integration"

This reverts commit ed54635510856cb8d556b7fc7288bb71a1617538.

boswelja committed:
Revert "Bump GRPCIO to support new integration"

This reverts commit ed54635510856cb8d556b7fc7288bb71a1617538.

boswelja committed:
Pin grpcio-reflection

boswelja committed:
Pin grpcio-reflection

boswelja commented:
I think I fixed it 🎊

boswelja committed:
Use a more random "address" for starlink tests
Using the default address was causing cases where it ran fine locally but not remotely (guess why)

boswelja committed:
Fix tests

boswelja committed:
Update .strict-typing

boswelja committed:
Run hassfest

slyglif commented:
Just curious if this is going to move forward or if the grpc conflict isn’t resolvable in a satisfactory way?

bdraco commented:
protobuf was unpinned on `dev` a week or so ago so there shouldn't be anything blocking this PR anymore

boswelja commented:
I'm curious too. According to CI, I had all issues fixed and this was ready to go.

If all is good, I can run hassfest and we can get this in asap

lyricnz commented:
I personally want a configurable poll interval before we merge this - last time I looked it was 1 second, even at 10 second, that's a bit too chatty: too much network, too many values.

MrConorAE commented:
The Starlink app uses a 1-second poll rate or thereabouts, so I think 1 second is an OK default for this. Having it configurable would be good though too.
------- Original Message -------
On Sunday, December 18th, 2022 at 15:47, Simon Roberts ***@***.***> wrote:


> I personally want a configurable poll interval before we merge this - last time I looked it was 1 second, even at 10 second, that's a bit too chatty: too much network, too many values.
> 

> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you commented.[blob:https://mail.proton.me/fe443af0-ce1a-4806-9fbd-183bfd160505]

lyricnz commented:
The starlink app is not open 24/7, and doesn't store all the results in a DB.  Hence configurable :)

lyricnz commented:
We seem to have two competing implementations for starlink integration?  How to decide which is best, and which is most likely to be maintained moving forwards?

MrConorAE commented:
Is there any chance of combining the two?
I'm currently using hastarlink on my HA instance, and it works great. How much do they differ?

boswelja commented:
Seems like the other integration has binary sensors, which I can't do in the same PR, and I've been waiting for this one to go in so I can build on it 🙂

boswelja commented:
On configurable poll time, that was what my original implementation had but I was advised to remove it (see earlier comments)

lyricnz commented:
I'm not sure I agree with 1 second as a "default/sane polling interval".  Unless you're actively using the visibility data in realtime (hardly a typical case), getting updates every 10 (or 60) seconds would be fine.

bdraco commented:
We generally don't merge PRs that have configurable polling.

We should choose the lowest polling interval the device can handle.  If users want to change it, they can disable polling and write an automation to update the entity instead.

In general, the shorted allowed polling interval is currently 5s

lyricnz commented:
THIS PR appears to have an interval of 1s?

https://github.com/home-assistant/core/pull/77091/files#diff-e5aeeff4aa5ef868fd44af40941290199700e0d75510fff523526361c9cd18ddR27
```update_interval=timedelta(seconds=1),```

bdraco commented:
> THIS PR appears to have an interval of 1s?
> 
> [#77091 (files)](https://github.com/home-assistant/core/pull/77091/files#diff-e5aeeff4aa5ef868fd44af40941290199700e0d75510fff523526361c9cd18ddR27) `update_interval=timedelta(seconds=1),`

I'm sure that will change before its merged.  It still needs code review now that its unblocked.

boswelja commented:
Polling interval of 1 second is intended, Dishy updates every second so that is "the lowest interval the device supports)

boswelja committed:
Merge branch 'dev' into feature/starlink-integration

boswelja committed:
Fix deprecated constants

zorrobyte commented:
I've been using https://github.com/archerne/hastarlink and it likes to timeout and not recover if my dishy goes offline and restarts. I have to do a reload manually. It'd be nice if one of these integrations could re-try automatically as I'm in an RV and turn off/on my starlink system often

boswelja committed:
Merge branch 'dev' into feature/starlink-integration

boswelja committed:
Extract some patchers for device found/no device found

boswelja committed:
Fix patchers

boswelja committed:
Bump starlink-grpc-core

boswelja commented:
This should be good for review again, I have increased code coverage for config_flow tests (though it doesn't look like that's an issue in CI anymore regardless), and made sure we are on the latest dependencies

bdraco commented on homeassistant/components/starlink/coordinator.py:
```diff
@@ -0,0 +1,38 @@
+        )
+
+    async def _async_update_data(self) -> StatusDict:
+        async with async_timeout.timeout(1):
```
```suggestion
        async with async_timeout.timeout(4):
```

1 seems like a low timeout here.  If the event loop is busy it going to fail.

boswelja replied:
Would it make more sense to move the timeout into the `add_executor_job` lambda in this case?

bdraco replied:
If the timeout is happening in the thread you'll effectively get the full timeout.  If you use `async_timeout` its going to include the time it takes for the event loop to pick up the job, hand it off to the executor and get the response back.

boswelja replied:
Ah ok, so not necessary here. Thanks!

bdraco commented on homeassistant/components/starlink/entity.py:
```diff
@@ -0,0 +1,70 @@
+        """Get a DeviceInfo for Dishy."""
+        config_url = f"http://{self.coordinator.channel_context.target.split(':')[0]}"
+        return DeviceInfo(
+            identifiers={
+                (DISHY_ID, self.coordinator.data["id"]),
+                (DISHY_HARDWARE, self.coordinator.data["hardware_version"]),
+            },
```
These need to be unique ids.

https://developers.home-assistant.io/docs/entity_registry_index?_highlight=acceptable#example-acceptable-sources-for-a-unique-id

boswelja replied:
I'll remove `hardware_version` since I'm not sure how unique that is, but `id` is definitely unique

bdraco commented on homeassistant/components/starlink/strings.json:
```diff
@@ -0,0 +1,17 @@
+{
+  "config": {
+    "abort": {
+      "already_configured": "Starlink terminal is already configured"
+    },
+    "error": {
+      "server_error": "Could not reach Starlink terminal."
```
Please use the standard `key::common` messages here

boswelja replied:
Marking as resolved so we only have one thread for this

bdraco commented on homeassistant/components/starlink/strings.json:
```diff
@@ -0,0 +1,17 @@
+{
+  "config": {
+    "abort": {
+      "already_configured": "Starlink terminal is already configured"
```
Please use the standard `key::common` messages here

bdraco replied:
By using the `key::common` strings we don't place additional work burden on translators

bdraco commented on tests/components/starlink/test_config_flow.py:
```diff
@@ -0,0 +1,50 @@
+        )
+        await hass.async_block_till_done()
+
+    assert result["type"] == data_entry_flow.FlowResultType.FORM
+    assert result["errors"]
+    hass.config_entries.flow.async_abort(result["flow_id"])
+    await hass.async_block_till_done()
```
Please finish all config flows tests to an ABORT to CREATE_ENTRY state.

boswelja replied:
I have created a test to make sure the flow is aborted when trying to create duplicates, if there's anything else we need to test please let me know 🙂

bdraco commented on homeassistant/components/starlink/config_flow.py:
```diff
@@ -0,0 +1,55 @@
+                lambda: get_id(context)
+            )
+            context.close()
+            return response
+        except GrpcError:
+            return None
```
```suggestion
        except GrpcError:
            return None
        else:
            return response
```

Please move code out of the `try` block that isn't expected to raise

bdraco commented on homeassistant/components/starlink/config_flow.py:
```diff
@@ -0,0 +1,55 @@
+        errors = {}
+        if user_input:
+            # Input validation. If everything looks good, create the entry
+            uid = await self.get_device_id(url=user_input[CONF_IP_ADDRESS])
+            if uid:
```
```suggestion
            if uid := await self.get_device_id(url=user_input[CONF_IP_ADDRESS]):
```

bdraco commented on homeassistant/components/starlink/__init__.py:
```diff
@@ -0,0 +1,35 @@
+
+    await coordinator.async_config_entry_first_refresh()
+
+    hass.data[DOMAIN] = {entry.entry_id: coordinator}
```
If they have multiple starlink integration this will override the data for the domain. use `setdefault` instead

bdraco commented on homeassistant/components/starlink/entity.py:
```diff
@@ -0,0 +1,70 @@
+        return DeviceInfo(
+            identifiers={
+                (DISHY_ID, self.coordinator.data["id"]),
+                (DISHY_HARDWARE, self.coordinator.data["hardware_version"]),
```
```suggestion
```

bdraco commented on homeassistant/components/starlink/entity.py:
```diff
@@ -0,0 +1,70 @@
+        config_url = f"http://{self.coordinator.channel_context.target.split(':')[0]}"
+        return DeviceInfo(
+            identifiers={
+                (DISHY_ID, self.coordinator.data["id"]),
```
```suggestion
                (DOMAIN, self.coordinator.data["id"]),
```

bdraco reviewed (dismissed):
Please see comments above

boswelja commented on homeassistant/components/starlink/sensor.py:
```diff
@@ -0,0 +1,79 @@
+        icon="mdi:upload",
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement=UnitOfDataRate.MEGABITS_PER_SECOND,
+        value_fn=lambda data: round(data["uplink_throughput_bps"] / 1000000),
+    ),
+    StarlinkSensorEntityDescription(
+        key="downlink_throughput",
+        name="Downlink throughput",
+        icon="mdi:download",
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement=UnitOfDataRate.MEGABITS_PER_SECOND,
+        value_fn=lambda data: round(data["downlink_throughput_bps"] / 1000000),
```
Fair enough, are there plans for pretty formatting data rates in UI? e.g. take a large bits/s and turn it into mbit/s, similar to how times work now

bdraco replied:
No sure when / if that will happen, but I'd move the data mutating to the lib in this case and wrap it in a dataclass so its easy to adjust later

boswelja committed:
Apply some changes from code review

boswelja committed:
Fix registering multiple Starlink integrations

boswelja committed:
Create a test to make sure the flow is aborted when trying to create duplicate entries

bdraco commented on tests/components/starlink/test_config_flow.py:
```diff
@@ -0,0 +1,80 @@
+        await hass.async_block_till_done()
+
+    assert result["type"] == data_entry_flow.FlowResultType.FORM
+    assert result["errors"]
+    hass.config_entries.flow.async_abort(result["flow_id"])
+    await hass.async_block_till_done()
```
We shouldn't abort the flow here.  We should validate they can still get to success state by entering the correct data after we get to the error state here.

ex the test is done when we reach a `data_entry_flow.FlowResultType.CREATE_ENTRY` or `data_entry_flow.FlowResultType.ABORT` state

boswelja replied:
I have updated the test, what do you think?

boswelja committed:
Fix up test_init

bdraco commented on homeassistant/components/starlink/strings.json:
```diff
@@ -0,0 +1,17 @@
+{
+  "config": {
+    "abort": {
+      "already_configured": "Starlink terminal is already configured"
+    },
+    "error": {
+      "server_error": "Could not reach Starlink terminal."
+    },
+    "step": {
+      "user": {
+        "data": {
+          "ip_address": "[%key:common::config_flow::data::ip%]"
+        }
+      }
+    }
+  }
+}
```
```suggestion
{
  "config": {
    "abort": {
      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]"
    },
    "error": {
      "cannot_connect": "[%key:common::config_flow::error::cannot_connect%]"
    },
    "step": {
      "user": {
        "data": {
          "ip_address": "[%key:common::config_flow::data::ip%]"
        }
      }
    }
  }
}
```

bdraco commented on homeassistant/components/starlink/config_flow.py:
```diff
@@ -0,0 +1,52 @@
+                    title="Starlink",
+                    data=user_input,
+                )
+            errors[CONF_IP_ADDRESS] = "server_error"
```
```suggestion
            errors[CONF_IP_ADDRESS] = "cannot_connect"
```

bdraco commented on homeassistant/components/starlink/translations/en.json:
```diff
@@ -0,0 +1,17 @@
+{
```
Regenerate this file with `python3 -m script.translations develop`

boswelja replied:
Is there somewhere in the documentation these strings are explained? I didn't even realize they were generated 😅

bdraco replied:
https://developers.home-assistant.io/blog/2020/04/10/translation-handling

bdraco replied:
https://github.com/home-assistant/architecture/blob/master/adr/0009-translations-2.0.md

boswelja committed:
Update homeassistant/components/starlink/strings.json

Co-authored-by: J. Nick Koston <nick@koston.org>

boswelja committed:
Update homeassistant/components/starlink/config_flow.py

Co-authored-by: J. Nick Koston <nick@koston.org>

boswelja committed:
Regenerate strings

bdraco commented on homeassistant/components/starlink/__init__.py:
```diff
@@ -0,0 +1,36 @@
+
+    await coordinator.async_config_entry_first_refresh()
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = coordinator
```
```suggestion
    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = coordinator
```

bdraco commented on homeassistant/components/starlink/entity.py:
```diff
@@ -0,0 +1,69 @@
+    @property
+    def device_info(self) -> DeviceInfo | None:
+        """Get a DeviceInfo for Dishy."""
+        config_url = f"http://{self.coordinator.channel_context.target.split(':')[0]}"
+        return DeviceInfo(
+            identifiers={
+                (DOMAIN, self.coordinator.data["id"]),
+            },
+            sw_version=self.coordinator.data["software_version"],
+            hw_version=self.coordinator.data["hardware_version"],
+            name="Starlink",
+            configuration_url=config_url,
+            manufacturer="SpaceX",
+            model="Starlink",
+        )
```
You can set this as `self._attr_device_info = ` in `__init__`

boswelja replied:
I guess initializing attributes in `__init__` where possible is generally preferred?

bdraco replied:
yes

boswelja committed:
Update test to make sure the user can still successfully configure the integration even when it fails once

boswelja committed:
Update homeassistant/components/starlink/__init__.py

Co-authored-by: J. Nick Koston <nick@koston.org>

boswelja committed:
Configure DeviceInfo in entity __init__

bdraco commented on homeassistant/components/starlink/config_flow.py:
```diff
@@ -0,0 +1,52 @@
+        """Get the device UID, or None if no device exists at the given URL."""
+        context = ChannelContext(target=url)
+        try:
+            response = await self.hass.async_add_executor_job(lambda: get_id(context))
```
```suggestion
            response = await self.hass.async_add_executor_job(get_id, context)
```

bdraco replied:
I don't think the `lambda` is needed here?

bdraco commented on homeassistant/components/starlink/coordinator.py:
```diff
@@ -0,0 +1,38 @@
+        async with async_timeout.timeout(4):
+            try:
+                status = await self.hass.async_add_executor_job(
+                    lambda: status_data(self.channel_context)
```
```suggestion
                   status_data, self.channel_context
```

Is the `lambda` needed here?

boswelja replied:
I thought so to pass variables, but it seems like that's not the case

boswelja committed:
Update homeassistant/components/starlink/coordinator.py

Co-authored-by: J. Nick Koston <nick@koston.org>

boswelja committed:
Update homeassistant/components/starlink/config_flow.py

Co-authored-by: J. Nick Koston <nick@koston.org>

boswelja committed:
Fix indent

boswelja committed:
Add missing fields to patchers

bdraco commented:
Looks like there is a test failure otherwise this looks good

I'll factory reset mine this weekend and try it out when I can afford to have it offline

boswelja commented:
Thanks for your time! I have fixed the tests locally, CI should run fine now

bdraco committed:
Merge branch 'dev' into feature/starlink-integration

bdraco commented:
Factory reset was a bit of a PITA but it works great!

```
2023-01-07 19:42:08.642 INFO (SyncWorker_6) [homeassistant.loader] Loaded starlink from homeassistant.components.starlink
2023-01-07 20:10:35.887 INFO (MainThread) [homeassistant.setup] Setting up starlink
2023-01-07 20:10:35.888 INFO (MainThread) [homeassistant.setup] Setup of domain starlink took 0.0 seconds
2023-01-07 20:10:35.907 INFO (MainThread) [homeassistant.components.sensor] Setting up sensor.starlink
2023-01-07 20:10:35.908 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new sensor.starlink entity: sensor.starlink_ping
2023-01-07 20:10:35.909 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new sensor.starlink entity: sensor.starlink_azimuth
2023-01-07 20:10:35.909 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new sensor.starlink entity: sensor.starlink_elevation
2023-01-07 20:10:35.910 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new sensor.starlink entity: sensor.starlink_uplink_throughput
2023-01-07 20:10:35.910 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new sensor.starlink entity: sensor.starlink_downlink_throughput
2023-01-07 20:10:35.911 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new sensor.starlink entity: sensor.starlink_last_boot_time
```

<img width="729" alt="Screenshot 2023-01-07 at 10 10 53 AM" src="https://user-images.githubusercontent.com/663432/211168733-985f94c1-31e9-413f-9f8b-65e8f6dee5f9.png">

bdraco commented:
Thanks @boswelja

lyricnz commented:
Works for me.  Glad polling is 5s and not 1s.

<img width="1562" alt="image" src="https://user-images.githubusercontent.com/122371/211176314-baa97f79-8449-45ee-b436-b58f7e7f3ccf.png">

rossfisherkc commented:
There's quite a few sensors that could be added, imo:
<img width="1213" alt="image" src="https://user-images.githubusercontent.com/113053906/211176359-c2a8c846-5060-496b-92b3-fdadf5e874f2.png">
(from the other integration)

I rely on the location, for example, to set my HA home location.
<img width="432" alt="image" src="https://user-images.githubusercontent.com/113053906/211176377-2ffa8243-718b-4969-98f3-af47857f9424.png">

Having access to, and tracking all of the metrics is pretty swank.

lyricnz commented:
Agree. I assume this merge is just the first step towards parity with the third-party repo https://github.com/archerne/hastarlink

rossfisherkc commented:
> Agree. I assume this merge is just the first step towards parity with the third-party repo https://github.com/archerne/hastarlink

Right! Getting things merged into mainline is hard enough, not complaining ;)

Great work getting this merged!

boswelja commented:
> Most people will think in mbit/sec
That's what I had it at, but we're not supposed to be modifying data from the Py library/API

> There's quite a few sensors that could be added
Guidelines say we can only add one platform at a time. You are showing binary sensors and buttons. I have a PR up for binary sensors now: https://github.com/home-assistant/core/pull/85409
I'll need to poke around in the response data to find fields for the remaining sensors (ping drop rate, lat/long etc), but again one thing at a time, I'd like to get binary sensors and buttons going first 🙂

lyricnz commented:
Is scaling/unit conversion really _modifying_?

boswelja commented:
Considering I was asked to stop doing that, yes

bdraco commented on homeassistant/components/starlink/entity.py:
```diff
@@ -0,0 +1,64 @@
+from .coordinator import StarlinkUpdateCoordinator
+
+
+@dataclass
+class StarlinkSensorEntityDescriptionMixin:
+    """Mixin for required keys."""
+
+    value_fn: Callable[[StatusDict], datetime | StateType]
+
+
+@dataclass
+class StarlinkSensorEntityDescription(
+    SensorEntityDescription, StarlinkSensorEntityDescriptionMixin
+):
+    """Describes a Starlink sensor entity."""
+
+
+class StarlinkSensorEntity(CoordinatorEntity[StarlinkUpdateCoordinator], SensorEntity):
```
I missed this on the first review but commented here: https://github.com/home-assistant/core/pull/85409#discussion_r1064096540

This should be split up into a `StarlinkSensorEntity` and a base class `StarlinkEntity` for the entity (the non-sensor) parts should live in `entity.py` and the sensor specific parts should move to `sensor.py`

MartinHjelmare commented on script/gen_requirements_all.py:
```diff
@@ -73,6 +73,7 @@
 # want to ensure we have wheels built.
```
This PR changed the permissions for this file to 644 from 755. This is a script and it should be executable.

MartinHjelmare replied:
Fixed in https://github.com/home-assistant/core/pull/85745.

bdraco replied:
That's a new one. Sorry I missed that. Thanks for fixing 👍

bdraco replied:
Figured out how I missed it....GitHub app doesn't show permission changes

bdraco replied:
I opened report #1964196 for the app

MartinHjelmare replied:
In the android app the indentation is completely missing for non changed lines. That's the show stopper for me with the app.

MartinHjelmare reviewed (commented):
Please address the comments in a new PR. Thanks!

boswelja commented on homeassistant/components/starlink/sensor.py:
```diff
@@ -0,0 +1,79 @@
+        device_class=SensorDeviceClass.TIMESTAMP,
+        entity_category=EntityCategory.DIAGNOSTIC,
+        value_fn=lambda data: datetime.now().astimezone()
+        - timedelta(seconds=data["uptime"]),
```
I haven't noticed any issues, but I also haven't run this for longer than maybe a minute at a time. I can leave it polling for a while and see if you suspect there might be an issue?
